### PR TITLE
fixing module.__file__ is None in py3

### LINKED ISSUE #4636: [bug] conanfile loader "crashes" because of missing __file__ attribute in Python 3.7(+)
Hi there,
I am using multiple Python versions (e.g. 3.6, 3.7). My primary Python version is 3.6.
But occationally I execute Python version 3.7. With 3.6 there where no problems. 3.7 does not work.

I have 2 projects. One (```pkg_a```) gets ```python_requires```d into the other (```pkg_b```).
pkg_b inherits from a python file that gets exported with pkg_a. Now Python 3.6 everything works fine. In Python 3.7 I get an exception.

Crawling through the conan code I identified (at least I think) the code where the solution lies.
At [this code line](https://github.com/conan-io/conan/blob/4fdebfb79ce12f956b4be512253e8479e9d6bb17/conans/client/loader.py#L265) conan can catch an exception if the python module has no ```__file__``` attribute.
But in Python 3.7 I get a ```TypeError``` exception instead of an ```AttributeError``` exception.

The [Python docs describe](https://docs.python.org/3.6/library/exceptions.html#AttributeError) ([see](https://docs.python.org/3.6/library/exceptions.html#AttributeError)) that there could be raised a ```TypeError``` exception.

In turn I modified the code line [265](https://github.com/conan-io/conan/blob/4fdebfb79ce12f956b4be512253e8479e9d6bb17/conans/client/loader.py#L265) in the conan [loader.py](https://github.com/conan-io/conan/blob/develop/conans/client/loader.py).

From:
```python
except AttributeError: # some module doesn't have __file__
    pass
```

To:
```python
except AttributeError: # some module doesn't have __file__
    pass
except TypeError: # Python 3.7: some module doesn't have __file__
    pass
```

And everything worked like before.
I do not know if this is a Python or Conan problem. But at the moment this is a real issue as noone with a version of Python 3.7+ can execute my recipe.

For reference I made a minimal example to track down the issue.
[py36_py37_exception.zip](https://github.com/conan-io/conan/files/2915266/py36_py37_exception.zip)

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Bugfix: Solve problem with loading recipe python files in Python 3.7 because of ``module.__file__ = None``
Docs: omit

closes #4636
close #4641

@PYVERS: Macos@py27, Windows@py36, Linux@py27, py34

### CODE REVIEW

File: conans/client/loader.py [hunk 1]
```diff
@@ -261,7 +261,10 @@ def _parse_conanfile(conan_file_path):
             module = sys.modules[added]
             if module:
                 try:
-                    folder = os.path.dirname(module.__file__)
+                    try:
+                        folder = os.path.dirname(module.__file__)
+                    except (AttributeError, TypeError):  # Namespace packages py3 module.__file__
+                        folder = module.__path__._path[0]
                 except AttributeError:  # some module doesn't have __file__
```
Comment (lasote) [inline conans/client/loader.py]: So, this except is for the `module.__path__._path[0]`? The comment is not very clear.

**Comment by lasote** [comment]
> That use-case makes sense with a python module installed in the system, and I'm not sure about those global/static variables inside a library,

That is valid and currently used. Pure python. About the global state, not so important but indeed a performance issue.

**Comment by jgsogo** [comment]
Yes, it is a valid use case, @lasote, but there are two things that are incompatible when loading dynamically python files:

 1. What it is here: import once and every recipe will use the same. It could make sense with a logger, output,... and that is the reason Conan is providing those utilities.
 2. The other perspective: **every recipe should behave the same regardless of the rest of the recipes**. Think about a `constants.py` file instead of a logger. This example is about two recipes following the same pattern:

    conanfile.py (this snippet is valid for `recipe1` and `recipe2`):

    ```python
    from constants import URL, DESC

    class MyLib(ConanFile):
       url = URL
       description = DESC
       exports = "constants.py"
    ```

    constants.py (accompaning `recipe1`)

    ```python
    URL = "http://library1"
    DESC = "This library implements..."
    ```

    constants.py (accompaning `recipe2`)

    ```python
    URL  = "http://library2
    DESC = "Functionality XXX"
    ```

    If this two recipes happen to be in the same _Conan instance_, the resulting behavior will be wrong, and it will depend on the order they are considered... and they would interact with a hook that could define a `constants.py` file too!

I think that both use cases are licit, but if I have to choose, **I prefer reproducibility and consistency over sharing a global variable** (and being afraid of not being using the one I pretended to use).

**Comment by memsharded** [comment]
The ``constants.py`` files accompanying each recipe will be put in the ``uuid1.constants`` and ``uuid2.constants`` different namespaces, and belonging exclusively to each recipe, not depending in the order, and being reproducible and independent of the order of evaluation.

The problem comes with modules that do not belong to recipes, like globally installed python libraries. These cannot be put under ``uuidX.something`` duplicated for every recipe in the graph.

**Comment by jgsogo** [comment]
> The `constants.py` files accompanying each recipe will be put in the `uuid1.constants` and `uuid2.constants` different namespaces, and belonging exclusively to each recipe, not depending in the order, and being reproducible and independent of the order of evaluation.

Yes, this PR is implementing that behavior too, you are true, it is doing the same I was proposing in #4664 and #4641 (plus the global thing).

Great! 👍

**Pull request merged**